File: utilities/src/main/kotlin/lunalauf/rms/utilities/network/client/Connector.java
package lunalauf.rms.utilities.network.client;

import javafx.concurrent.Task;
import lunalauf.rms.utilities.network.util.ConnectionInitiationHelper;
import lunalauf.rms.utilities.network.util.PortProvider;

import java.io.IOException;
import java.net.InetSocketAddress;
import java.net.Socket;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;
import java.util.function.Consumer;

public class Connector {

    private static final int timeout = 200; // ms
    private final ExecutorService executor;
    private ConnectionTask currentTask = null;

    public Connector() {
        this.executor = Executors.newSingleThreadExecutor();
    }

    public void connect(String host, int port, Consumer<Client> onConnected, Runnable onAborted) {
        cancel();
        currentTask = new ConnectionTask(host, port);
        currentTask.setOnSucceeded(event -> {
            try {
                Client client = currentTask.get();
                if (client != null)
                    onConnected.accept(client);
                else
                    onAborted.run();
            } catch (Exception e) {
                onAborted.run();
            }
        });
        currentTask.setOnCancelled(event -> onAborted.run());
        currentTask.setOnFailed(event -> onAborted.run());
        executor.execute(currentTask);
    }

    public void cancel() {
        if (currentTask != null)
            currentTask.cancel();
    }

    public void shutdown() {
        cancel();
        try {
            executor.shutdown();
            if (!executor.awaitTermination(3, TimeUnit.SECONDS))
                executor.shutdownNow();
        } catch (InterruptedException ignored) {
        }
    }

    private class ConnectionTask extends Task<Client> {
        private final String host;
        private final int port;

        private ConnectionTask(String host, int port) {
            this.host = host;
            this.port = port;
        }

        @Override
        protected Client call() throws Exception {
            while (true) {
                if (isCancelled())
                    return null;

                if (this.port > 0 && this.port <= 65535) {
                    try {
                        Socket socket = new Socket();
                        socket.connect(new InetSocketAddress(host, port), timeout * 2);
                        Client client = new Client(socket);
                        if (initialCommunicationTest(client))
                            return client;
                        else {
                            client.close();
                            throw new Exception("Communication test failed");
                        }
                    } catch (IOException ignored) {
                    }
                } else {
                    for (int prefPort : PortProvider.getPreferredPorts()) {
                        if (isCancelled())
                            return null;
                        try {
                            Socket socket = new Socket();
                            socket.connect(new InetSocketAddress(host, prefPort), timeout);
                            Client client = new Client(socket);
                            if (initialCommunicationTest(client))
                                return client;
                            else
                                client.close();
                        } catch (Exception ignored) {
                        }
                    }
                }
            }
        }
    }

    private boolean initialCommunicationTest(Client client) throws Exception {
        client.setTimeout(5000);
        try{
            String synMessage = ConnectionInitiationHelper.getSynMessage();
            String expectedAckMessage = ConnectionInitiationHelper.getAckMessage(synMessage);

            client.send(synMessage);
            String ackMessage = client.receive();

            if (!expectedAckMessage.equals(ackMessage))
                return false;

            String synAckMessage = ConnectionInitiationHelper.getAckMessage(ackMessage);
            client.send(synAckMessage);
            return true;
        } catch (Exception e) {
            return false;
        } finally {
            client.resetTimeout();
        }
    }

}

File: utilities/src/main/kotlin/lunalauf/rms/utilities/network/client/Connector.kt
package lunalauf.rms.utilities.network.client

import kotlinx.coroutines.*
import lunalauf.rms.utilities.network.util.ConnectionInitiationHelper
import lunalauf.rms.utilities.network.util.PortProvider
import lunalauf.rms.utilities.network.util.Service
import java.io.IOException
import java.net.InetSocketAddress
import java.net.Socket

class Connector : Service<Connector.Input, Connector.Result>(Dispatchers.IO) {
    companion object {
        private const val TIMEOUT = 200 // ms
        private const val TIMEOUT_COMMUNICATION_TEST = 5000 // ms
    }

    override fun CoroutineScope.run(input: Input): Result {
        while (true) {
            if (!isActive) return Result.Aborted

            if (input.port in 1..65535) {
                try {
                    val socket = Socket()
                    socket.connect(InetSocketAddress(input.host, input.port), TIMEOUT * 2)
                    val client = Client(socket)
                    return if (initialCommunicationTest(client)) {
                        logger.info("Connected to port: {}", client.port)
                        Result.Connected(client)
                    } else {
                        client.close()
                        Result.Failed("Communication test failed")
                    }
                } catch (_: Exception) {
                }
            } else {
                for (prefPort in PortProvider.getPreferredPorts()) {
                    if (!isActive) return Result.Aborted

                    try {
                        val socket = Socket()
                        socket.connect(InetSocketAddress(input.host, prefPort), TIMEOUT)
                        val client = Client(socket)
                        return if (initialCommunicationTest(client)) {
                            logger.info("Connected to port: {}", client.port)
                            Result.Connected(client)
                        } else {
                            client.close()
                            Result.Failed("Communication test failed")
                        }
                    } catch (_: Exception) {
                    }
                }
                logger.info("Could no connect to any preferred port. Retrying...")
            }
        }
    }

    @Throws(IOException::class)
    private fun initialCommunicationTest(client: Client): Boolean {
        client.setTimeout(TIMEOUT_COMMUNICATION_TEST)
        return try {
            val synMessage: String = ConnectionInitiationHelper.synMessage
            val expectedAckMessage: String = ConnectionInitiationHelper.getAckMessage(synMessage)
            client.send(synMessage)
            val ackMessage = client.receive()
            if (expectedAckMessage != ackMessage) {
                logger.warn("Communication test failed")
                return false
            }
            val synAckMessage: String = ConnectionInitiationHelper.getAckMessage(ackMessage)
            client.send(synAckMessage)
            logger.info("Communication test successful")
            true
        } catch (e: Exception) {
            logger.warn("Communication test failed due to an exception", e)
            false
        } finally {
            client.resetTimeout()
        }
    }

    data class Input(
        val host: String,
        val port: Int
    )

    sealed class Result {
        data object Aborted : Result()
        data class Failed(val reason: String) : Result()
        data class Connected(val client: Client) : Result()
    }
}
